flowmachine/flowmachine/core/server/query_schemas/modal_location.py

```diff
@@ -0,0 +1,57 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from marshmallow import Schema, fields, post_load
+from marshmallow.validate import OneOf, Length
+from marshmallow_oneofschema import OneOfSchema
+
+from .base_exposed_query import BaseExposedQuery
+from .custom_fields import AggregationUnit, SubscriberSubset
+from .daily_location import DailyLocationSchema, DailyLocationExposed
+
+
+class InputToModalLocationSchema(OneOfSchema):
```

Comment:
I don't think there's actually a better way to do this, but I don't like it much.. Primarily because it adds hoops to creating a new thing that you can feed into modal locations.

Reply:
I used to share this sentiment. I agree it feels a little clunky at first, but the more I think about it the more I actually like it because it creates a meaningful abstraction that states: "these are precisely the types of things that can be fed into `ModalLocation`".

I don't think it creates extra hoops actually, because the only difference to if this class didn't exist is that now you need to add a new class to the list`InputToModalLocationSchema.type_schemas`, and otherwise you'd need to add it to `FlowmachineQuerySchema.type_schemas`. But apart from that the process of adding something new that you can feed to modal locations is identical.

Reply:
I _think_ what I'm wondering is why one can't have all the schemas inherit from the base one, deal with the type field bit by marking it as `dump_only` or something like that, and then define the things that can be accepted in terms of inheritance and, in openapi world, `allOf`, e.g. you have a 'meta' schema/mixin scenario, so if your schema wants to be accepted as input to modal locations it includes the `ReferenceLocation` schema in the bases and modal location then `InputToModalLocationSchema(FlowmachineQuerySchema, ReferenceLocation)`, or something like that.

But I think it just bugs me because it introduces a weird bidirectional inheritance situation, where the superclass is in charge of declaring child classes :P

Reply:
Interesting points, I'll mull over this when I have a bit more time and headspace. :)

My first approach was actually to use `dump_only` but I couldn't quite get it to work (which doesn't mean it's impossible, of course) and even the incomplete attempts didn't feel very tidy. But there may well be a better way of doing things along the lines you suggested. Happy to revisit after this is merged if we find a better approach.